File: sdks/react-native/fantasma-react-native/android/src/main/java/com/fantasma/reactnative/FantasmaReactNativeModule.kt
package com.fantasma.reactnative

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.fantasma.sdk.FantasmaConfig
import com.fantasma.sdk.FantasmaException
import com.fantasma.sdk.reactnative.FantasmaReactNativeBridge
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

public class FantasmaReactNativeModule(
    reactContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(reactContext) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val bridge = FantasmaReactNativeBridge(reactContext.applicationContext)

    override fun getName(): String = "FantasmaReactNative"

    @ReactMethod
    public fun open(
        serverUrl: String,
        writeKey: String,
        promise: Promise,
    ) {
        scope.launch {
            runCatching {
                bridge.open(
                    FantasmaConfig(
                        serverUrl = serverUrl,
                        writeKey = writeKey,
                    ),
                )
            }.fold(
                onSuccess = { promise.resolve(null) },
                onFailure = { promise.rejectBridgeError(it) },
            )
        }
    }

    @ReactMethod
    public fun track(
        eventName: String,
        promise: Promise,
    ) {
        scope.launch {
            runCatching { bridge.track(eventName) }.fold(
                onSuccess = { promise.resolve(null) },
                onFailure = { promise.rejectBridgeError(it) },
            )
        }
    }

    @ReactMethod
    public fun flush(promise: Promise) {
        scope.launch {
            runCatching { bridge.flush() }.fold(
                onSuccess = { promise.resolve(null) },
                onFailure = { promise.rejectBridgeError(it) },
            )
        }
    }

    @ReactMethod
    public fun clear(promise: Promise) {
        scope.launch {
            runCatching { bridge.clear() }.fold(
                onSuccess = { promise.resolve(null) },
                onFailure = { promise.rejectBridgeError(it) },
            )
        }
    }

    @ReactMethod
    public fun close(promise: Promise) {
        scope.launch {
            runCatching { bridge.close() }.fold(
                onSuccess = { promise.resolve(null) },
                onFailure = { promise.rejectBridgeError(it) },
            )
        }
    }

    override fun invalidate() {
        super.invalidate()
        scope.cancel()
    }
}

private fun Promise.rejectBridgeError(error: Throwable) {
    when (error) {
        FantasmaException.InvalidWriteKey -> reject(
            "invalid_write_key",
            error.message,
            error,
        )
        FantasmaException.UnsupportedServerUrl -> reject(
            "unsupported_server_url",
            error.message,
            error,
        )
        FantasmaException.InvalidEventName -> reject(
            "invalid_event_name",
            error.message,
            error,
        )
        FantasmaException.EncodingFailed -> reject(
            "encoding_failed",
            error.message,
            error,
        )
        FantasmaException.InvalidResponse -> reject(
            "invalid_response",
            error.message,
            error,
        )
        FantasmaException.UploadFailed -> reject(
            "upload_failed",
            error.message,
            error,
        )
        is FantasmaException.StorageFailure -> reject(
            "storage_failure",
            error.detail,
            error,
        )
        FantasmaException.ClosedClient -> reject(
            "closed_client",
            error.message,
            error,
        )
    }
}

File: sdks/react-native/fantasma-react-native/ios/FantasmaReactNativeModule.swift
import Foundation

#if canImport(React)
import React

@objc(FantasmaReactNative)
final class FantasmaReactNativeModule: NSObject {
    private let bridge = FantasmaReactNativeBridge.shared

    @objc
    static func requiresMainQueueSetup() -> Bool {
        false
    }

    @objc(open:writeKey:resolver:rejecter:)
    func open(
        serverURL: String,
        writeKey: String,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                guard let url = URL(string: serverURL) else {
                    throw FantasmaError.unsupportedServerURL
                }
                try await bridge.open(serverURL: url, writeKey: writeKey)
                resolve(nil)
            } catch {
                rejectFantasmaError(error, reject: reject)
            }
        }
    }

    @objc(track:resolver:rejecter:)
    func track(
        eventName: String,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                try await bridge.track(eventName)
                resolve(nil)
            } catch {
                rejectFantasmaError(error, reject: reject)
            }
        }
    }

    @objc(flush:rejecter:)
    func flush(
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                try await bridge.flush()
                resolve(nil)
            } catch {
                rejectFantasmaError(error, reject: reject)
            }
        }
    }

    @objc(clear:rejecter:)
    func clear(
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                try await bridge.clear()
                resolve(nil)
            } catch {
                rejectFantasmaError(error, reject: reject)
            }
        }
    }

    @objc(close:rejecter:)
    func close(
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            await bridge.close()
            resolve(nil)
        }
    }

    private func rejectFantasmaError(
        _ error: Error,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        if let fantasmaError = error as? FantasmaError {
            switch fantasmaError {
            case .invalidWriteKey:
                reject("invalid_write_key", fantasmaError.localizedDescription, fantasmaError)
            case .unsupportedServerURL:
                reject("unsupported_server_url", fantasmaError.localizedDescription, fantasmaError)
            case .invalidEventName:
                reject("invalid_event_name", fantasmaError.localizedDescription, fantasmaError)
            case .encodingFailed:
                reject("encoding_failed", fantasmaError.localizedDescription, fantasmaError)
            case .invalidResponse:
                reject("invalid_response", fantasmaError.localizedDescription, fantasmaError)
            case .uploadFailed:
                reject("upload_failed", fantasmaError.localizedDescription, fantasmaError)
            case .storageFailure(let detail):
                reject("storage_failure", detail, fantasmaError)
            case .notConfigured:
                reject("closed_client", "Fantasma client has already been closed.", fantasmaError)
            }
            return
        }

        reject("storage_failure", error.localizedDescription, error)
    }
}
#endif
